### ISSUE TITLE
Errors during action handlers

### ISSUE THREAD

**Issue by mbfisher** [yahoo/dispatchr#46]
TL;DR If an error is thrown whilst handling an action the `Dispatcher` will not allow any further actions to be dispatched.

The logic that causes this is, I think, contained within `Dispatcher::dispatch`:
```
Dispatcher.prototype.dispatch = function dispatch(actionName, payload) {
    if (this.currentAction) {
        throw new Error('Cannot call dispatch while another dispatch is executing. Attempted to execute \'' + actionName + '\' but \'' + this.currentAction.name + '\' is already executing.');
    }
    ...
    this.currentAction = new Action(actionName, payload);
    ....
    this.currentAction.execute(handlerFns);
    debug('finished ' + this.currentAction.name);
    this.currentAction = null;
```
If an error is thrown during `Action::execute`, `this.currentAction` is never reset to `null` and so further calls to `dispatch` throw the 'Cannot call dispatch' error.

Using `try` around the `Action::execute` call prevents the 'Cannot call dispatch' error from being thrown under legitimate circumstances, in fact there is a test case ensuring that `dispatch()` does not swallow errors thrown by Store handlers.

Given that the current behaviour renders an app unusable, what are the suggestions for handling errors in a Fluxible app?

M

**Comment by tildedave** [comment]
Handling this better would be very nice - I have to try/catch while debugging because exceptions (from my code) get swallowed by the Dispatchr error handling mechanism.  Guidance on best practices would be appreciated!

**Comment by mridgway** [comment]
Ok, so most of this stems from the fact that we're not resetting `currentAction` after a thrown exception. Perhaps we can add a `try/catch/finally` to the `execute` method to ensure that it gets reset.

**Comment by mbfisher** [comment]
@mridgway this is hopefully what's achieved in #47.

M

**Comment by mridgway** [comment]
I'm not sure that we need to differentiate the internal errors with the store's errors in this case. Both should still throw, but they should be able to be handled by the caller to dispatch. Also, I think in your implementation the `this.currentAction` wouldn't get reset since throw would exit the flow.

Here is my proposal for fixing this: https://github.com/yahoo/dispatchr/commit/66f5355dc7c7333ef85a6e40aeec053830f2740f

**Comment by mbfisher** [comment]
Yup, you're right! That looks spot on.

M
